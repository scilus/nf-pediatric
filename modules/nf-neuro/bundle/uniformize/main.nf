process BUNDLE_UNIFORMIZE {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.0.sif':
        'scilus/scilus:2.0.2' }"

    input:
    tuple val(meta), path(bundles), path(centroids)

    output:
    tuple val(meta), path("*_uniformized.trk"), emit: bundles
    path "versions.yml"                       , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def method = task.ext.method ? "--${task.ext.method}"  : "--auto"
    def swap = task.ext.swap ? "--swap" : ""

    """
    bundles=(${bundles.join(" ")})

    if [[ -f "$centroids" ]]; then
        centroids=(${centroids.join(" ")})
    fi

    for index in \${!bundles[@]};
        do \
        bname=\$(basename \${bundles[index]} .trk)
        if [[ -f "$centroids" ]]; then
            option="--centroid \${centroids[index]}"
        else
            option="$method"
        fi
        scil_bundle_uniformize_endpoints.py \${bundles[index]} ${prefix}__\${bname}_uniformized.trk\
            \${option}\
            $swap -f
    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    scil_bundle_uniformize_endpoints.py -h

    for bundles in ${bundles};
        do \
        bname=\$(basename \${bundles} .trk)
        touch ${prefix}__\${bname}_uniformized.trk
    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
