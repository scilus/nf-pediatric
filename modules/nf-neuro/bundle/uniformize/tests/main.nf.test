nextflow_process {

    name "Test Process BUNDLE_UNIFORMIZE"
    script "../main.nf"
    process "BUNDLE_UNIFORMIZE"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "bundle"
    tag "bundle/uniformize"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "bundles.zip" , "tractometry.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("bundle - uniformize") {
        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        bundles: it.simpleName == "bundles"
                        tractometry: it.simpleName == "tractometry"
                    }
                input[0] = ch_split_test_data.bundles.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/fibercup_atlas/subj_1/*.trk"),
                        []
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("bundle - uniformize_axis") {

        config "./nextflow_axis.config"

        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        bundles: it.simpleName == "bundles"
                        tractometry: it.simpleName == "tractometry"
                    }
                input[0] = ch_split_test_data.bundles.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/fibercup_atlas/subj_1/bundle_0.trk"),
                        []
                    ]
                }
                """
            }
        }
    then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("bundle - uniformize_centroid") {

        config "./nextflow_centroid.config"

        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        bundles: it.simpleName == "bundles"
                        tractometry: it.simpleName == "tractometry"
                    }
                input[0] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/IFGWM.trk"),
                        file("\${test_data_directory}/IFGWM_uni_c_10.trk")
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
