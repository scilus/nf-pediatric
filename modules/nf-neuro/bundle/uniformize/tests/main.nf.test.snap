{
    "bundle - uniformize": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__bundle_0_uniformized.trk:md5,cd25919f690cdec5874abaa2dd960213",
                            "test__bundle_1_uniformized.trk:md5,b159c84734897c9c34bcf654ab78e60a",
                            "test__bundle_2_uniformized.trk:md5,eb78c3ba6090d33b901fc28152679bce",
                            "test__bundle_3_uniformized.trk:md5,8877161b9aebaa6b12a4d7a8f8410407",
                            "test__bundle_4_uniformized.trk:md5,d7bb41aa3f9cc0d3596c738bd67fe4e0",
                            "test__bundle_5_uniformized.trk:md5,e2ff89549b0cc1c84bc54abfc665542d",
                            "test__bundle_6_uniformized.trk:md5,6659d3fd4c3a34ac75ece918f20e9ebf"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,f2a592bbc31c9c2393317ab72bcabe8c"
                ],
                "bundles": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__bundle_0_uniformized.trk:md5,cd25919f690cdec5874abaa2dd960213",
                            "test__bundle_1_uniformized.trk:md5,b159c84734897c9c34bcf654ab78e60a",
                            "test__bundle_2_uniformized.trk:md5,eb78c3ba6090d33b901fc28152679bce",
                            "test__bundle_3_uniformized.trk:md5,8877161b9aebaa6b12a4d7a8f8410407",
                            "test__bundle_4_uniformized.trk:md5,d7bb41aa3f9cc0d3596c738bd67fe4e0",
                            "test__bundle_5_uniformized.trk:md5,e2ff89549b0cc1c84bc54abfc665542d",
                            "test__bundle_6_uniformized.trk:md5,6659d3fd4c3a34ac75ece918f20e9ebf"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f2a592bbc31c9c2393317ab72bcabe8c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-22T20:18:24.833877"
    },
    "bundle - uniformize_axis": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bundle_0_uniformized.trk:md5,cd25919f690cdec5874abaa2dd960213"
                    ]
                ],
                "1": [
                    "versions.yml:md5,f2a592bbc31c9c2393317ab72bcabe8c"
                ],
                "bundles": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__bundle_0_uniformized.trk:md5,cd25919f690cdec5874abaa2dd960213"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f2a592bbc31c9c2393317ab72bcabe8c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-22T20:19:03.734244"
    },
    "bundle - uniformize_centroid": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__IFGWM_uniformized.trk:md5,ad2c69fa67ca1fd1269b0db7dff4fe07"
                    ]
                ],
                "1": [
                    "versions.yml:md5,f2a592bbc31c9c2393317ab72bcabe8c"
                ],
                "bundles": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__IFGWM_uniformized.trk:md5,ad2c69fa67ca1fd1269b0db7dff4fe07"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f2a592bbc31c9c2393317ab72bcabe8c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-22T20:14:52.97367"
    }
}